sycl/doc/design/KernelFusionJIT.md

```diff
@@ -0,0 +1,173 @@
+
+### Event Handling
+
+The calls to `queue::submit` as well as to `queue::ext_codeplay_complete_fusion` need to return an event to allow synchronization with the execution of submitted kernels or the fused kernel.
+
+Returning the event associated with the individual kernels (`ExecCGCommand`) from `queue::submit` while in fusion mode would mean that these events become useless if the fusion mode is later on completed successfully because the individual kernels are removed from the graph and never executed in this case.
+
+To overcome this problem, a call to `queue::submit` while in fusion mode will instead return the event associated with the `KernelFusionCommand`.
+This event will always remain valid and allow for synchronization, independent of whether the fusion mode was terminated through early cancellation, `ext_codeplay_cancel_fusion`, or `ext_codeplay_complete_fusion`, as the KernelFusionCommand in all these scenarios has event dependencies on all relevant commands.
+
+Another important scenario is a call to `ext_codeplay_complete_fusion` after the fusion for this queue has been aborted early (see section on [Synchronization Behavior](#synchronization-behavior)).
+In this case, the individual kernels and the KernelFusionCommand could have been enqueued and even executed long ago, but the call to `complete_fusion` still needs to return a valid event for synchronization.
+To handle this case, the KernelFusionCommand for each queue remains in the fusion map until the next `ext_codeplay_start_fusion` on the same queue and does not undergo the cleanup process until its status has been set to `DELETE` by `ext_codeplay_start_fusion`.
+This way, the lifetime of the event associated with the KernelFusionCommand is extended such that it is still valid for synchronization when returned from a later call to `ext_codeplay_complete_fusion`.
+
+Note that even though the KernelFusionCommand is associated with a particular queue and context, the associated event does not have an underlying PI event, i.e., it is essentially a host event.
+
+
+## Fusion Process
+
+To fuse kernels, a small LLVM based JIT is responsible to perform the following tasks:
+
+ - Load input modules in LLVM IR format (translate SPIR-V module to LLVM IR for now)
+ - Fuse kernels in a new module
+ - Perform fusion specific optimization
+ - Finalize for the target (emit SPIR-V module for now)
+
+The pipeline is triggered by the SYCL runtime by calling the `fuseKernels` function and its result is then injected into the runtime as a kernel bundle.
+
+The fusion specific optimization is composed of new LLVM passes we wrote for this task.
+
+The rest of the code is just plumbing code to build the pipeline. The SPIR-V loading and emission are done by the LLVM-SPIRV-Translator.
+
+As this is an LLVM based JIT, we can integrate it in various ways.
```

Comment:
Can we add more details on that or this would be too low-level for this design document?

Reply:
It is legacy from the original document. Nothing to expand on.